### PR TITLE
Fixed the sample method. 

### PR DESCRIPTION
**Pull request by odinsbane**
Colormap.sample would go out of bounds. It was also broken for odd h values. [Issue #750](https://github.com/scenerygraphics/scenery/issues/750)

Made a hot fix for generating an image form the ColormapEditor the image was missing 10px so depending on how large the editor was displayed the resulting Colormap always returned black. [Issue #757](https://github.com/scenerygraphics/scenery/issues/757)

If this looks ok I can improve it.

### CODE REVIEW

**Comment by skalarproduktraum** [comment]
Looks good to me, thank you, @odinsbane! I think it'd only need some formatting cleanup, and the -10 removed, right?

File: src/main/kotlin/graphics/scenery/volumes/ColormapPanel.kt [hunk 1]
```diff
@@ -304,8 +304,8 @@ class ColormapPanel(val target:Volume?): JPanel() {
 
         internal fun toImage(): BufferedImage {
             val rec: Rectangle = this.bounds
-            val bufferedImage = BufferedImage(rec.width, rec.height, BufferedImage.TYPE_INT_ARGB)
-            paintBackgroundGradient(colorPoints.sortedBy { it.position }, bufferedImage.graphics as Graphics2D)
+            val bufferedImage = BufferedImage(rec.width, rec.height - 10, BufferedImage.TYPE_INT_ARGB)
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/volumes/ColormapPanel.kt:307]: When the sampling is fixed, the -10 can go away again I think

File: src/main/kotlin/graphics/scenery/volumes/Colormap.kt [hunk 2]
```diff
@@ -27,30 +27,36 @@ class Colormap(val buffer: ByteBuffer, val width: Int, val height: Int) {
...
-        val c1 = Vector4f(ub[0].toFloat() / 255.0f, ub[1].toFloat() / 255.0f, ub[2].toFloat() / 255.0f, ub[3].toFloat() / 255.0f)
-        val c2 = Vector4f(ub[4].toFloat() / 255.0f, ub[5].toFloat() / 255.0f, ub[6].toFloat() / 255.0f, ub[7].toFloat() / 255.0f)
+        val b = buffer.duplicate()
+        b.position(globalOffset + previous * 4);
+        val color = ByteArray(4)
+        b.get(color);
+        //Add to "Image" utility class?
+        val c1 = Vector4f( color[0].toUByte().toFloat() / 255f, color[1].toUByte().toFloat() / 255f, color[2].toUByte().toFloat() / 255f, color[3].toUByte().toFloat() / 255f)
+
+        if( bufferPosition > previous ){
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/volumes/Colormap.kt:51]: Formatting should be adjusted to the general format of the project (sorry we don't have a style file :-/)

File: src/main/kotlin/graphics/scenery/volumes/Colormap.kt [hunk 3]
```diff
@@ -27,30 +27,36 @@ class Colormap(val buffer: ByteBuffer, val width: Int, val height: Int) {
...
+            //interpolate fraction part.
+            b.get(color);
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/volumes/Colormap.kt:53]: Semicolon is not necessary

File: src/main/kotlin/graphics/scenery/volumes/Colormap.kt [hunk 4]
```diff
@@ -27,30 +27,36 @@ class Colormap(val buffer: ByteBuffer, val width: Int, val height: Int) {
...
+            val c2 = Vector4f( color[0].toUByte().toFloat() / 255f, color[1].toUByte().toFloat() / 255f, color[2].toUByte().toFloat() / 255f, color[3].toUByte().toFloat() / 255f)
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/volumes/Colormap.kt:54]: Should not have space after the brackets. Sorry again for not having a style file.

File: src/main/kotlin/graphics/scenery/volumes/Colormap.kt [hunk 5]
```diff
@@ -27,30 +27,36 @@ class Colormap(val buffer: ByteBuffer, val width: Int, val height: Int) {
...
+            return c1.lerp(c2, bufferPosition - previous.toFloat())
+        } else{
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/volumes/Colormap.kt:56]: Space missing before `{`.

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
Just some minor comments. Thanks @odinsbane 👍

**Comment by odinsbane** [comment]
Great! I don't have a kotlin development environment so I need to pay
attention to the formatting a little.

The "-10" is in the draw background and it is at the point of creating the
buffered image. It should probably be fixed so the background image is
filled and the buffered image is the correct size. Without the -10 the
markers get painted over. It might take a little refactoring to remove.

File: src/main/kotlin/graphics/scenery/volumes/ColormapPanel.kt [hunk 6]
```diff
@@ -324,8 +322,11 @@ class ColormapPanel(val target:Volume?): JPanel() {
             val h = height
             val pointList = colorPoints.sortedBy { it.position }
 
+
             // background Gradient
-            paintBackgroundGradient(pointList, g2d)
+            //paintBackgroundGradient(pointList, g2d)
```
Comment (moreApi) [inline src/main/kotlin/graphics/scenery/volumes/ColormapPanel.kt:327]: remove or uncomment

File: src/main/kotlin/graphics/scenery/volumes/Colormap.kt [hunk 7]
```diff
@@ -27,30 +27,44 @@ class Colormap(val buffer: ByteBuffer, val width: Int, val height: Int) {
...
-        val c2 = Vector4f(ub[4].toFloat() / 255.0f, ub[5].toFloat() / 255.0f, ub[6].toFloat() / 255.0f, ub[7].toFloat() / 255.0f)
+        // The number of bytes per pixel are fixed at 4.
+        val globalOffset = width * band * 4
 
-        return c1.lerp(c2, bufferPosition - previous.toFloat())
+        val b = buffer.duplicate()
+        b.position(globalOffset + previous * 4)
+        val color = ByteArray(4)
+        b.get(color)
+        // Add to "Image" utility class?
```
Comment (moreApi) [inline src/main/kotlin/graphics/scenery/volumes/Colormap.kt:47]: what to do with such comments?

**Review by moreApi — CHANGES_REQUESTED** [review]
lgtm execpt those two comments. I like that magic numbers get replaced with propper variable names :)

File: src/main/kotlin/graphics/scenery/volumes/Colormap.kt (hunk 7 above)
Comment (skalarproduktraum) [reply src/main/kotlin/graphics/scenery/volumes/Colormap.kt:54]: Added the requested utility function

File: src/main/kotlin/graphics/scenery/volumes/ColormapPanel.kt (hunk 6 above)
Comment (skalarproduktraum) [reply src/main/kotlin/graphics/scenery/volumes/ColormapPanel.kt]: Removed 👍

**Review by moreApi — APPROVED** [review]

**Review by skalarproduktraum — APPROVED** [review]

**Comment by skalarproduktraum** [comment]
Thanks @odinsbane! I took the liberty of fixing the remaining issues myself, I hope you don't mind.

**Pull request merged**